### PR TITLE
Make GraphQL Client work for Rails 5 projects

### PR DESCRIPTION
**Pull request by dewski**
When trying to use the current GraphQL Client on Rails 5 applications it would fail to boot using the Railtie which ties everything together.

To fix it:

- Initialize the GraphQL::Client in the context of the Railtie module, it was being initialized without the required `schema` kwarg. This would raise.

https://github.com/github/graphql-client/blob/773b8260f5af0307a5a6400e81f74e1c60261071/lib/graphql/client.rb#L86-L88

- Detect and use [Erubi](https://github.com/jeremyevans/erubi) when Erubis is not present.
- Provide standard query execution / context helpers for `ActionController::Base`.
- Don't inject the current value of `Rails.application.config.graphql.client` at initialization time of the initial `ViewModule` in the Railtie or any time a module is loaded. Instead use a method inside of `ViewModule` that references `Rails.application.config.graphql.client` so the application can configure their client in an initializer.

### TODO

- [x] Update documentation on how to use the Railtie & controller helpers. (ie: `viewer: current_user`)
- [x] Update documentation about making the client accessible at `Rails.application.config.graphql.client`.

Open to feedback on how to best let the application let GraphQL Client know about the client to use. Another approach is to just lookup the schema under a pre determined module name, but that approach isn't very flexible whereas applications can configure it however they like and set it to `Rails.application.config.graphql.client`.

/cc @josh @arthurschreiber @nakajima

### CODE REVIEW

**Comment by antn** [comment]
Hey folks, I just ran into this when trying to use graphql-client for a rails 5 project i'm working on. Do y'all have the bandwidth to get this shipped? No worries if not!

**Comment by dewski** [comment]
> Hey folks, I just ran into this when trying to use graphql-client for a rails 5 project i'm working on. Do y'all have the bandwidth to get this shipped? No worries if not!

I can circle back around to this now that there is more demand 🙂

Curious if @josh or @arthurschreiber are 👍 with re-working the documentation to require people to setup `Rails.application.config.graphql.client`. I could also create a Rails generator to help assist with the installation / migration.

**Comment by dewski** [comment]
Have updated this PR to finish support for a more automated setup:

- Added a Rails generator `rails generate graphql_client:install` with documentation.
- Injected the `def graphql_context` method in ApplicationController to allow apps to configure the context, also documented.
- Updated the README to help with how to use inside views in a Rails project.

Installing GraphQL::Client and running:

```
$ rails generate graphql:install
$ rails generate graphql_client:install
```

Will have you setup with a working application.

The tests aren't cooperating at all, need to look into that further.

Thoughts @josh @arthurschreiber?

File: lib/graphql/client/controller_helpers.rb [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+      #              Client.parse("query { version }") returns a definition.
+      # variables - Optional set of variables to use during the operation.
+      #             (default: {})
+      #
+      # Returns a structured query result or raises if the request failed.
+      def graphql_query(definition, variables = {})
+        response = graphql_client.query(definition, variables: variables, context: graphql_context)
+
+        if response.errors.any?
+          raise GraphQL::Client::QueryError.new(response.errors[:data].join(", "))
```
Comment (josh) [inline lib/graphql/client/controller_helpers.rb]: I've still never felt great about the raising on top level errors behavior—but I guess @github has been using it so maybe we should just go with it.

File: lib/graphql/client/view_module.rb [hunk 2]
```diff
@@ -17,7 +17,10 @@ class Client
     #   Views::Users::Show::UserFragment
     #
     module ViewModule
-      attr_accessor :client
+      def client
+        @client ||= Rails.application.config.graphql.client
+      end
+      attr_writer :client
```
Comment (josh) [inline lib/graphql/client/view_module.rb]: I'm unsure of this change will interact with reloading. We're memorizing the client reference here but what if we want to reload this reference? Also it seems like the `attr_writer` behavior is broken. Anything that attempts to assign a custom `client` reference won't be propagated to any modules.

File: README.md [hunk 3]
```diff
@@ -153,6 +171,59 @@ These `<%graphql` sections are simply ignored at runtime but make their definiti
...
+After you run the [provided Rails generator](#configuration) it will add a `graphql_context` method for you to use to pass any additional context for your queries.
+
+For example:
+
+```ruby
+class ApplicationController < ActionController::Base
+  def graphql_context
+    {
+      viewer: current_user,
+      tenant: current_tenant,
```
Comment (josh) [inline README.md]: I don't think setting AR objects in context is a best practice any new applications should be using. Can we prompt passing some kind of auth/api token string here?

File: lib/graphql/client/view_module.rb (hunk 2 above)
Comment (dewski) [reply lib/graphql/client/view_module.rb]: Maybe we just need to be more opioniated and require the client to be set in an initializer and only reference `Rails.application.config.graphql.client`?

**Comment by eileencodes** [comment]
Hello! The erubis -> erubi is now a blocker for us on upgrading GitHub to Rails 5.2. Can I help move this PR along?

**Comment by rlgreen91** [comment]
Hmm, I'm wondering about this too.  Not critical atm but will be eventually.

**Comment by dewski** [comment]
@josh I've updated the `graphql_context` example and how the `GraphQL::Client::ViewModule` knows about the configured GraphQL client. Is this good to go now?

File: lib/graphql/client/view_module.rb [hunk 4]
```diff
@@ -105,11 +103,10 @@ def load_module(name)
         query, lineno = ViewModule.extract_graphql_section(contents)
         return unless query
 
-        mod = client.parse(query, path, lineno)
+        mod = Rails.application.config.graphql.client.parse(query, path, lineno)
```
[the remark this answers is not in the document]
Comment (dewski) [reply lib/graphql/client/view_module.rb]: Is there an implementation you'd feel comfortable moving forward with?

**Comment by josh** [comment]
I cherry picked out some erubis changes here. `master` should be green on rails-52.

**Comment by rlgreen91** [comment]
@josh I just updated to the latest version of the gem, running on Rails 5.2.  I'm still getting the same error: 

```
ArgumentError:
  missing keyword: schema
```
Here's a snapshot of my `application.rb`:

...
```
require "graphql/client/railtie"
require "graphql/client/http"

# Require the gems listed in Gemfile, including any gems
# you've limited to :test, :development, or :production.
Bundler.require(*Rails.groups)

module VsaDriver
  class Application < Rails::Application
    # Initialize configuration defaults for originally generated Rails version.
    config.load_defaults 5.2

    # Settings in config/environments/* take precedence over those specified here.
    # Application configuration can go into files in config/initializers
    # -- all .rb files in that directory are automatically loaded after loading
    # the framework and any gems in your application.

    # Only loads a smaller set of middleware suitable for API only apps.
    # Middleware like session, flash, cookies can be added back manually.
    # Skip views, helpers and assets when generating a new resource.
    config.api_only = true
    config.active_job.queue_adapter = :sidekiq
  end

  HTTP = GraphQL::Client::HTTP.new("#{ENV.fetch('EXELON_HOST')}/graphql") do
    def headers(context)
      { "X-SITE-ID": "1"}
    end
  end

  Schema = GraphQL::Client.load_schema(HTTP)

  Client = GraphQL::Client.new(schema: Schema, execute: HTTP)

  Application.config.graphql.client = Client
end
```

**Comment by d12** [comment]
Any movement on this? It's blocking one of my GraphQL projects on a Rails 5 app.

**Comment by tobsch** [comment]
+1

**Comment by antn** [comment]
@dewski @josh 👋 it looks like this PR was mistakenly merged and then reverted in https://github.com/github/graphql-client/pull/162. i believe the railtie in master is still broken on rails 5/6. is it possible to wrap up this PR and get the railtie working? this is kind of an old PR now, so totally understandable if you have other priorities and can't get to this, just figured i'd check in :)

**Pull request merged**